### DISCUSSION TITLE
Device agnostic code

### DISCUSSION THREAD

**Discussion by ra9hur** [mrdbourke/pytorch-deep-learning#282] [Q&A] [answered]
At the outset, thanks for a very well-structured course. I am coming from tensorflow background and this course is helping me a lot in picking up PyTorch related concepts quickly !!

Working with devices in tensorflow is very intuitive and we just have to select device for the session OR specify distributed strategy and things work out of the box.

Have two questions related to this.

1. Any thoughts on why we have to choose device explicitly in PyTorch ?

2. Choosing device for the model makes sense since model has to run on a given device. 

2a. Why should we choose device for data ?

2b. If we choose GPU:1 for data and and gpu:0 for model (separate devices for model and data), it should introduce latency and will be noticeable for bigger datasets. Why is this option given ? 

The design should have been that data / model to reside on the same device to speed up things.
Are these options given to choose depending on RAM size of devices, size of datasets and estimated RAM size that could be required to train the model ?

Sure, I am missing something very trivial. Need your inputs.

**Comment by mrdbourke** [comment, marked as answer]
Hi @ra9hur,

Good questions!

> 1. Any thoughts on why we have to choose device explicitly in PyTorch ?

This is for maximum customizability of the code.

And the ability to choose which device the computation happens on.

E.g. some computation may be better on GPU, some better on CPU.

> 2. Choosing device for the model makes sense since model has to run on a given device.

> 2a. Why should we choose device for data ?

Model + data have to be on the same device for computation to happen.

If a model is on GPU and data on CPU, an error will occur (the data isn't stored in the same place).

> 2b. If we choose GPU:1 for data and and gpu:0 for model (separate devices for model and data), it should introduce latency and will be noticeable for bigger datasets. Why is this option given ?

This is correct.

For this, you'll likely want to use libraries that allow multi-GPU usage, such as:

* Hugging Face Accelerate - https://huggingface.co/docs/accelerate/index
* Torch FullyShardedDataParallel - https://pytorch.org/docs/stable/fsdp.html

**Reply by ra9hur** [reply]
Hi @mrdbourke, thanks for clarifying, helped a lot !!
